### DISCUSSION TITLE
[Feature]: Android TV app/interface

### DISCUSSION THREAD

**Discussion by jagjordi** [immich-app/immich#1614] [Feature Request]
### Feature detail

As suggested by @alextran1502 in #1182 I am opening this issue to keep track of the idea of having a Android TV app or some sort of interface to be able to view immich photos/albums on a TV. 

### Platform

Web

**Comment by alextran1502** [comment]
Reference implementation information
https://medium.com/@polincgames/flutter-android-tv-apps-checklist-part-1-basic-requirements-83e0adc2e144

**Comment by EnochPrime** [comment]
If may request here, one feature with a cast/TV interface that would be awesome, would be the ability to display exif data and/or other tags. It's near impossible to find a photo program that does that (at least according to my father).

**Reply by threehappypenguins** [reply]
Yes, yes, yes. A million times yes. I have scanned photos that I would like to be able to display the captions; stuff like that. Can't do that on my TV with Plex. I would just get rid of Plex completely if there were an Immich Android TV app (I have an Nvidia Shield TV).

**Comment by ikslavok** [comment]
Is this going anywhere? Why is this marked as done, if it's not done?

**Comment by Pheggas** [comment]
I don't think this is has high priority. Doubt it will be done by this (or even next) year.

**Comment by threehappypenguins** [comment]
Bump. I know this is old and doesn't seem to have priority, but now that Immich is an import for a read-only library, I'm one step closer to actually using Immich. After I read you can import, I installed it again today and it works wonderfully. So the next two major things for me are being able to import a second library (I keep pictures and videos separate) to the same user, and having an Android TV app (because we like to view everything as a family).

**Comment by briandipalma** [comment]
The ability to use it as a screensaver would be nice. We use Google photos for that currently. After 5 minutes of inactivity on the home screen we start to see random photos from specified Google photos albums.

**Reply by fivestones** [reply]
I second this

**Reply by akohlsmith** [reply]
I third this - a roku TV screensaver which displayed images with matching selected tags or albums would be A+

**Comment by el-fredo** [comment]
Do you really need a dedicated Google TV / Android TV app? In my opinion, a cast-feature would also do the job. The Google Photos app offers this feature, and I use it a lot to show photos to friends on my TV screen.

**Reply by Pheggas** [reply]
This topic would need solid discussion with developers. Alex said there's no cast feature YET. It is QOL feature as you can cast your screen and do it this way.

**Reply by threehappypenguins** [reply]
Casting may do the job, but most times I don't really want to use my phone, and my kids also like to be able to look and not all of them have a device. Right now, we use Plex, but it would be nice to just use Immich for everything. My in-laws also use Plex right now, and they don't really understand casting. We also share our library with them so that they can view our family albums. Can't do that with casting either.

**Reply by fivestones** [reply]
If this were to have functionality to run as the screensaver on an android tv (or the Chromecast with Google TV), cast ability wouldn't be enough.

I mean, I would really appreciate both--sometimes I'm on my phone and want to cast what I'm looking at to the TV. Other times I just want the inactive TV to start showing photos/videos as a slideshow screensaver.

**Reply by pedropombeiro** [reply]
I'm currently using the Plex Media Player on a Raspberry Pi so that my parents-in-law can use the TV remote to navigate family photos. They wouldn't be able to manage casting photos to the TV, as much as I see this covering the needs of 80% of the users.

Just my 2 cents.

**Comment by jagjordi** [comment]
In my view casting is sub-optimal. It requires the phone and the quality is not the best. An app would be much better.

I have sideloaded the immich apk into my android tv and for the most part it works quite good (with a mouse) but the interface with the remote is not good.

I think it might be possible to make it work with minimal work but I have no experience with that so I can't tell for sure

________________________________

**Reply by jrasm91** [reply]
I think there are two different feature requests here - (1) android tv support via cast and (2) android tv support via native app. Both have pros/cons. I'm not sure if one is easier to implement than the other. Either way I don't think they need to be mutually exclusive.

**Reply by jrasm91** [reply]
Stuff like a screensaver probably isn't possible without an app though.

**Comment by Peda1996** [comment]
Why not use a minidlna for this, as the folder structure is already useable?

**Reply by Pheggas** [reply]
There are already multiple ways how to show photos on Android TV. We're talking here about QOL feature, not "normal" feature

**Comment by edouardaihua** [comment]
Would also love to have a Cast feature or a Google TV app. We often watch pictures with my elderly family members on TV, and the Google Photos cast feature is very useful for that. At my place, i can indeed use another app and DLNA but at their place i can't .

**Comment by fivestones** [comment]
I'm not sure if it should be a separate request, but as far as I know, to use an app as a screen saver on android TV it can't just be any app but has to be built specifically to have screen saver functionality. I think the original request here is for an immich app that is usable with a remote on Android TV (which I would also like to have) but a second, related, (but separate?) request would be for a screen saver app which cycles through immich photos automatically as a screen saver. Options to have it use a certain album or to cycle through the search results of a certain search would be nice too.

**Comment by jjspierx** [comment]
Ideally, there would be an Immich app for EVERY thing, but realistically, that is not going to happen, at least not anytime soon.  A cast feature, would be a great stop gap that could work on multiple devices (Android TV, Roku, Firestick, etc...)

**Comment by fivestones** [comment]
If I'm not mistaken, a cast feature for Chromecast/Android TV and one for Roku (and probably for Firestick?) would all have to be created independently as they all have their own methods of making a cast feature work.

Once the slideshow feature is a little more mature, it should be very little effort to port the Immich android app to be a screensaver app on Android TV.

Creating a casting function for the immich mobile apps should be doable but more work as far as I can tell.

**Comment by aisbergde** [comment]
I tried transfer to chromecast:
- on PC using the stream function of chrome browser, 
- on Android using the stream function of the whole display in the Google home app. 
 
Somehow it works. But "Google Photo" streaming experience is better. Are there other ideas how to show immich assets on Chromecast? We are using this often for common viewing.

Slideshows on Chromecast and on Chrome Nest Hub are also running all the time in our home. It would be great to use immich as source, but not only Google.

**Reply by briandipalma** [reply]
I'm not sure there is much that can be done for the Nest Hub but an Android TV screensaver based on Immich pictures would be awesome.

**Reply by aisbergde** [reply]
simple Chromecast support would be fine.

**Reply by fivestones** [reply]
> an Android TV screensaver based on Immich pictures would be awesome

I think it might be two different features? But I'd love to have both.

**Reply by aisbergde** [reply]
of course I think, these are two different things. But the original #2575 was closed and commented as

> Duplicate of #1614

I don't agree, that this a duplicate. But what should I do? That's why it is mixed up here.

I would prefer to reopen the other issue.

**Reply by fivestones** [reply]
Hmmm. You're right, those aren't duplicates.

**Comment by giejay** [comment]
Just to let you know I'm currently working on a Android TV solution. First basic stuff like fetching and showing albums and a slider for pictures/videos works. Will try to include options for Screensaver and casting (but I don't think casting requires a native app). Will share some screenshots this week.

**Reply by threehappypenguins** [reply]
YES. Thank you! Is there also a way to show metadata (EXIF, IPTC, XMP, etc)? Specifically captions? There is no feasible solution for this for a TV app (or any other app). Not Plex, not Jellyfin, and not Emby. Emby comes the closest where you can edit the metadata, but you can't actually display it.

**Reply by giejay** [reply]
Seeing as it is available in the API response (api/album/{id}): 
```
"exifInfo": {
"make": "Google",
"model": "Pixel 6",
"exifImageWidth": 3072,
"exifImageHeight": 4080,
"fileSizeInByte": 3977014,
"orientation": "1",
"dateTimeOriginal": "2023-09-15T14:50:53.843Z",
"modifyDate": "2023-09-15T14:50:53.843Z",
"timeZone": "UTC+2",
"lensModel": "Pixel 6 back camera 6.81mm f/1.85",
"fNumber": 1.9,
"focalLength": 6.81,
"iso": 307,
"exposureTime": "1/106",
"latitude": 53.xxxx,
"longitude": 5.xxxx,
"city": "x",
"state": "Noardeast-Fryslân, Friesland",
"country": "Netherlands",
"description": "",
"projectionType": null
}
```
I don't see why not! Just need to integrate it properly design wise. I will add it to the backlog;) 

I am focusing on albums only for now, just showing all "assets" unordered can be added of course, but not sure if this is going to be used a lot.

**Reply by fivestones** [reply]
Amazing! Thanks for doing this!

**Reply by giejay** [reply]
I promised to share some screenshots: https://giejay-immich.duckdns.org/share/KBiPdwQfrLRWFb9W0CTOhpi4qK0PE_omUWAArFWnnuea6LXuAhsSI_cPi1UH4zpJ104

Working:
- Login screen to set hostname + API key
- Demo environment if you want to try it before adding credentials
- Album fetching + Lazy loading
- Assets in Album fetching + Lazy loading
- Showing the picture of a selected asset/album as the background
- Slideshow of the assets including video support 
- Playing the slideshow with 2.5 second interval

Working on:
- Screensaver support, including the filtering op albums to show in the screensaver

On the backlog:
- Setting the slideshow interval
- Searching for and in albums
- Casting
- Showing Exif data
- Improve error handling
- Dependency injection with Hilt/Dagger

**Reply by fivestones** [reply]
I know people don't say this any more, but you're the bomb.

**Reply by Pheggas** [reply]
Let Giejay cook 🔥

**Comment by ekdahl** [comment]
I guess another option would be if Immich would be added as a source in the Android TV app _Photo Gallery and Screensaver_, which I currently use as TV screensaver and to view albums from Google Photos and local SMB server.
https://photoscreensaver.furnaghan.com/

**Comment by giejay** [comment]
I have published a first alpha/beta version to the Play Store, which can be installed by joining the test release:  https://play.google.com/apps/testing/nl.giejay.android.tv.immich. Curious to receive your feedback! Still a lot to do, but you should be able to at least see all your albums and watch the photos/videos inside them. 

In the settings, you can set the Immich TV app as the screensaver of Android TV. The screensaver will simply pick the latest updated album and show the pictures in that album. I'm working on making the albums selectable and randomising the photos which are shown.

**Reply by pedropombeiro** [reply]
I joined the beta program, but it seems it is not compatible with my TV box (Android 11) 😞:

<img width="461" alt="image" src="https://github.com/immich-app/immich/assets/138074/49886b47-b5dc-4b70-bcbd-5f78b4336a24">

**Reply by giejay** [reply]
What's the exact model? Is it a native Android/Google TV box?

**Reply by tenfire** [reply]
I have the same problem with my NVIDIA Shield Pro 2019

**Reply by pedropombeiro** [reply]
@giejay this is a Technicolor uzx4020nps ([review article](https://www.xavierstuder.com/2020/06/le-test-du-nouveau-boitier-tv-android-de-net-la-carte-de-la-simplicite/) in French).

**Reply by giejay** [reply]
Strange, I also have the Shield Pro 2019 but it's installing just fine. Just like my Sony Bravia, Mi TV and a Mi Box. Will check if it's country related.

**Reply by jagjordi** [reply]
Do you have a github repo to track issues regarding the TV app?

**Reply by fivestones** [reply]
And is this app you are making open source?

**Reply by giejay** [reply]
I will publish it to Github once I feel comfortable with the code (need to clean it up). I might already create a repo to start tracking issues, good idea.

**Reply by tenfire** [reply]
The review seems to be completed, the link you posted before the edit is working for me now.

**Reply by giejay** [reply]
Ah great, I see a green flag indeed. Didn't get a mail yet so thanks for the notification :) I will revert the link back to the original open testing url link.

@ the people who mailed me, please use the new link to join the open test: https://play.google.com/apps/testing/nl.giejay.android.tv.immich. I will close the "closed test".

**Comment by ikslavok** [comment]
Just installed it on local brand TV with android v8.0 I tried both demo and my vault.
The interface looks really good. I like how selected photo is the background. I also like how fast it works on demo vault.

Bugs I noticed:
- videos are slow or if longer they just crash the app or don't load.
- Can't use local ip, which should be a common use case
- Initial viewing of photos is just so slow. Maybe there could be automatic indexing on first few photos of hovered album.

**Reply by giejay** [reply]
I will do some debugging with larger videos, I did not notice it in the demo environment (did you?).

Not sure why local IP would be an issue, http://192.168.2.x should just work, but maybe its restrictive to HTTPS, I will check!

Last point, do you mean in the album details view, or when playing the slideshow? What's your download/upload speed? I will try to mimick slower environments. Have gigabit myself so didn't notice any slowness when loading videos or photos. Would also be fixed for you if local IP would work of course.

**Reply by pedropombeiro** [reply]
One thing I noticed when loading a large album (almost 1000 pictures/videos) is that the background picture kept changing while the page loaded. It seemed to track the loading picture, and even though I wasn't interacting with the app, it kept changing the background picture (around once each second). This went on for around 10 seconds.

**Reply by ikslavok** [reply]
> I will do some debugging with larger videos, I did not notice it in the demo environment (did you?).

It crashed 30℅ into video. But I tried again and your demo env worked smooth.

> Not sure why local IP would be an issue, http://192.168.2.x should just work, but maybe its restrictive to HTTPS, I will check!

I have it on non standard port. Does it acept port as :xxxx or did you set it to specific port?

> Last point, do you mean in the album details view, or when playing the slideshow? What's your download/upload speed? I will try to mimick slower environments. Have gigabit myself so didn't notice any slowness when loading videos or photos. Would also be fixed for you if local IP woul

Both. Details is lil bit better. I have 80/20 Mbit so you might say its a slow network. Local host is definitely a solution.

**Reply by giejay** [reply]
Version 1.9 should give you the option to use: http://local IP:2138 for example. I will look into the crashing of the videos.

**Reply by giejay** [reply]
@pedropombeiro  Strange. I will investigate, thanks for the feedback!

**Reply by jagjordi** [reply]
Really amazing job @giejay . I highly encourage you to set up some more organized way to track the changes like a empty github repo or something.
For now I noticed some problem when playing landcape video after a portrait one, the rotation and scaling gets messed up somehow. it looks to me like the landscape video gets rotated and somehow the aspect ratio becomes the opposite as it should be. I am not sure what kind of data/info you need to be able to replicate/investigate that/ please let me know and I will try to provide it

**Reply by giejay** [reply]
If you could provide me a landscape and portrait video so I can reproduce it, that would be great! Please create a ticket on the new GitHub repo I shared, so we can track it.

**Comment by men1al** [comment]
Just had a look and it seems very promising, thanks for your work !
Tested with an account with a single album, photos and short videos. It worked great

I noticed, only direct albums on the API account are listed. It would be great to see the shared albums as well

**Reply by giejay** [reply]
Shared albums are visible in release 1.10

**Comment by giejay** [comment]
I'm releasing version 1.9 which adds/fixes the following:
- Enable HTTP Immich instances
- Adding shared albums
- Fix a NPE/crash when exifinfo is missing
- Randomize the order of the screensaver
- Fix other crashes I saw in Firebase Crashlytics

**Comment by giejay** [comment]
I opened a repository and will track changes in the changelog ticket. 

https://github.com/giejay/Immich-Android-TV

Feel free to open bug/feature request issues.

**Reply by giejay** [reply]
I have pushed the code as well

**Reply by MaBeniu** [reply]
Just to share here as well if you are not able to set screensaver most probably your TV Android OS version has restriction to not allow overriding Ambient Mode into custom screensaver. 
It is still possible to use screensaver but not straight forward. 
Workaround. 
Enable Dev mode on TV and run command:
```
adb shell settings put secure screensaver_components nl.giejay.android.tv.immich/.playback.ScreenSaverService
```
Otherwise I am very satisfied with the speed and quality. The only wish would be an option to view full libraries of specific user

**Reply by giejay** [reply]
You mean view all the photos instead of only the albums? That will come up in one of the next releases

**Reply by matticas** [reply]
Thanks for this, just started trying it now!  So far so good!

**Reply by palves** [reply]
OOC, why a new program from scratch instead of reusing the Android/phone app with interface tweaks?

**Reply by giejay** [reply]
Android TV requires a completely different library set (leanback), so that would be a major rewrite and also increase the regular app size a lot.

**Reply by palves** [reply]
You would be able to reuse components and logic like the timeline viewer, etc., just like Google themselves explain:

"If you have an existing Android app, adding Android TV support lets you design a user interface for TV while re-using your existing app architecture."

https://developer.android.com/training/tv/start/start

The fact that it's possible to sideload regular android apps on ATV shows that a lot of things should be reusable.

I don't think that necessarily it would increase the regular app size at all.  There could be different builds from the same codebase with different pieces.

**Reply by palves** [reply]
I looked a little, and the Immich android app is based on Flutter.  From articles like:

  https://medium.com/@polincgames/flutter-android-tv-devices-c53bb25200ee  (first of a series)

and 

 https://www.reddit.com/r/FlutterDev/comments/tcfgcq/lessons_learned_releasing_my_app_for_android_tv/

It seems like it's doable to port a flutter app to Android TV.

**Reply by giejay** [reply]
Thanks for the insights! One of the reasons I picked up this project is wanting to learn Kotlin. So for now, I will keep focusing on the standalone app. Maybe in the future, when I want to learn Flutter/Dart, I will have a go;)

**Comment by OGTK423** [comment]
Getting a "Could not fetch items! Status code: 401" after entering host name and API key on the setup screen. Web and app access currently setup and working.

**Reply by matticas** [reply]
Did you include the http & port number in the host?

http://192.168.0.x:2283/   ?

**Reply by OGTK423** [reply]
Yes. I can confirm both the host URL and API key are both entered accurately.

**Reply by matticas** [reply]
Sorry i'm out of ideas.  Maybe worth jumping on the Immich Android TV git and logging it there?

https://github.com/giejay/Immich-Android-TV

**Reply by hasecilu** [reply]
I had that problem yesterday when I just uploaded +600 photos and my server was quite busy, today tried again and working as expected.

**Comment by DeltaTango69** [comment]
Hi, nice project - is ist also planned to have the nomal tímeline view? I am not using any albums, exclusively only the timeline :-)

**Reply by giejay** [reply]
The photos tab will show all photos/videos uploaded, sorted by date, is that what you want?

**Reply by DeltaTango69** [reply]
I haven't installed the app yet. I've only read the information about it here so far and it seemed that only albums are supported. But if the Timeline View is already implemented, all the better : -) I have a Samsung TV and would like to use the app. Unfortunately, it only works with additional hardware running Android TV. The following features would be great for me:
- Playing photos and videos one after the other in chronological order (from old to new) from a certain image/video
- Support for 360° photos and videos
- Automatic transcoding of videos to the TV's maximum resolution of 4K (Insta 360° videos have a native resolution of 5.7k)

As I said, these are all just wishful thinking - in the hope that one day it will be possible :-)

**Reply by OGTK423** [reply]
Samsung's run Tizen TV OS which is a totally different operating system than Android TV.

**Reply by DeltaTango69** [reply]
Yes of course, therefore I need another hardware on android basis - any suggestions for an android mediaplayer?

**Reply by OGTK423** [reply]
Shield TV is the way to go for 99% of users.

**Comment by oekomat** [comment]
Hello, I think this is the right place for my question. I have the google tv 4k. I installed the app from the playstore. I can't get a connection to my self-hosted docker immich. The error is: Could not fetch data prom api. response: Unable to parse TLSpacket header.
I have no idea.

**Reply by giejay** [reply]
Hi, regarding the place, the better place is here: https://github.com/giejay/Immich-Android-TV

For your issue. Did you disable ssl verification when setting up the server? You can do it in settings - server - set up using API key.

**Reply by oekomat** [reply]
With a special variable? My configuration with docker:

docker run
  -d
  --name='immich'
  --net='br0'
  --ip='192.168.10.64'
  -e TZ="Europe/Berlin"
  -e HOST_OS="Unraid"
  -e HOST_HOSTNAME="Tower"
  -e HOST_CONTAINERNAME="immich"
  -e 'TCP_PORT_8080'='8080'
  -e 'DB_HOSTNAME'='192.168.10.20'
  -e 'DB_USERNAME'='postgres'
  -e 'DB_PASSWORD'='---'
  -e 'DB_DATABASE_NAME'='postgres'
  -e 'REDIS_HOSTNAME'='192.168.10.20'
  -e 'DB_PORT'='5432'
  -e 'REDIS_PORT'='6379'
  -e 'REDIS_PASSWORD'=''
  -e 'MACHINE_LEARNING_GPU_ACCELERATION'=''
  -e 'MACHINE_LEARNING_WORKERS'='1'
  -e 'MACHINE_LEARNING_WORKER_TIMEOUT'='120'
  -e 'EXTERNAL_PATH'='/mnt/user/Multimedia/xyz/'
  -e 'PUID'='99'
  -e 'PGID'='100'
  -e 'UMASK'='022'
  -l net.unraid.docker.managed=dockerman
  -l net.unraid.docker.webui='http://[IP]:[PORT:8080]'
  -l net.unraid.docker.icon='https://raw.githubusercontent.com/imagegenius/templates/main/unraid/img/immich.png'
  -v '/mnt/user/Multimedia/xyz/':'/photos':'rw'
  -v '/mnt/user/Multimedia/xyz/import/':'/import':'ro'
  -v '/mnt/user/appdata/immich':'/config':'rw' 'ghcr.io/imagegenius/immich'

**Reply by giejay** [reply]
No, inside the Android tv app.

**Reply by oekomat** [reply]
on my google tv 4k i have the actually immich app (1.29). I have no settings for ssl connections.

**Reply by oekomat** [reply]
sorry, my fail. it works. i have try to set with user and password. Now i have the newest version 1.32 too.

**Comment by radh21301** [comment]
I just came across this blog by ente photos. I am no software developer but I thought this might be of use to someone looking to help with the cast feature.

https://ente.io/blog/introducing-cast/

**Comment by bo0tzz** [comment]
Since @giejay has now built this as an awesome community project (thank you!), I'll close this feature request.

**Reply by Glideh** [reply]
I would have thought a common effort between the mobile Android and TV Android to be interesting.
Especially in terms of API & UI design implementation.
@giejay does a great work but he actually redoes many of the same things as in the mobile app.

Maybe I'm wrong, that was my two cents.
